qiskit/transpiler/target.py

```diff
@@ -0,0 +1,652 @@
+from qiskit.transpiler.timing_constraints import TimingConstraints
+
+logger = logging.getLogger(__name__)
+
+
+class InstructionProperties:
+    """A representation of the properties of a gate implementation.
+
+    This class provides the optional properties that a backend can provide
+    about an instruction. These represent the set that the transpiler can
+    currently work with if present. However, if your backend provides additional
+    properties for instructions you should subclass this to add additional
+    custom attributes for those custom/additional properties by the backend.
+    """
+
+    __slots__ = ("length", "error", "calibration")
+
+    def __init__(
+        self,
+        length: float = None,
+        error: float = None,
+        calibration: Union["Schedule", "ScheduleBlock"] = None,
+    ):
+        """Create a new ``InstructionProperties`` object
+
+        Args:
+            length: The duration, in seconds, of the instruction on the
+                specified set of qubits
+            error: The average error rate for the instruction on the specified
+                set of qubits.
+            calibration: The pulse representation of the instruction
+        """
+        self.length = length
+        self.error = error
+        self.calibration = calibration
+
+    def __repr__(self):
+        return (
+            f"InstructionProperties(length={self.length}, error={self.error}"
+            f", calibration={self.calibration})"
+        )
+
+
+class Target(Mapping):
+    """
+    A description of gates on a backend. It exists around a central mapping of
+    :class:`~qiskit.circuit.Instruction` objects to their properties on the
+    device. As a basic example, lets assume your device is two qubits, supports
+    :class:`~qiskit.circuit.library.UGate` on both qubits and
+    :class:`~qiskit.circuit.library.CXGate` in both directions you would create
+    the target like::
+
+        from qiskit.transpiler import Target, InstructionProperties
+        from qiskit.circuit.library import UGate, CXGate
+        from qiskit.circuit import Parameter
+
+        gmap = Target()
+        theta = Parameter('theta')
+        phi = Parameter('phi')
+        lam = Parameter('lambda')
+        u_props = {
+            (0,): InstructionProperties(length=5.23e-8, error=0.00038115),
+            (1,): InstructionProperties(length=4.52e-8, error=0.00032115),
+        }
+        gmap.add_instruction(UGate(theta, phi, lam), u_props)
+        cx_props = {
+            (0,1): InstructionProperties(length=5.23e-7, error=0.00098115),
+            (1,0): InstructionProperties(length=4.52e-7, error=0.00132115),
+        }
+        gmap.add_instruction(CXGate(), cx_props)
+
+    The intent of the ``Target`` object is to inform Qiskit's compiler about
+    the constraints of a particular backend so the compiler can compile an
+    input circuit to something that works and is optimized for a device.
```

Comment:
I have some concerns about the definition of `Target` and what it actually represents. At the top of the Target class it says:

> A description of gates on a backend.

But I read in a different comment that you wrote:

> As for the qubit properties used by the transpiler, yeah ideally we'd put those in the target. I want the target to be the representation of a backend for the transpiler (as called by the preset pass managers). Nothing in transpile() today takes T1/T2 as a parameter that I'm aware of. But, I think we can expand on it over time in a non-breaking way if needed. So for example if in 0.20.0 we add a pass to the default pass managers that can use t1 we should expand the target to have an interface for it. 

I actually agree with the first definition and disagree with the second. To me a backend has some qubits, and has some instructions. In fact I think "target" is exactly what I have been calling "ISA" all this time and added a bunch of github labels for. 

So I think at a high-level the backend should be organized into its qubits and its instructions. A compiler will usually use the info in the instructions to infer everything like gates, durations, errors, calibrations.
But if it needs T1, it can just read it from the qubit properties. I don't see why we need to encapsulate all info useful for compilation in "target". And perhaps the word "target" is confusing then. Perhaps something like "isa" or "instruction_set" is more explicit. 

But we may be thinking about this differently so I'd like to hear your thoughts.

(to be precise I think there are other things about a backend that don't fit into either its instruction_set or its qubits, like its hamiltonian which is useful for pulse simulation (perhaps even compilation if the compiler does some smart pulse finding)).

Reply:
Agreed. Properties such as T1 and T2 are arguments for the optional hardware agnostic optimization pass. Signature of transpile function should be flexible enough to take these noise options (perhaps some container), e.g. `transpile(target: Target, **kwargs)`. If we need to pack everything in `Target`, I wonder how we can set a scheduling option ("alap"/"asap"/None) as a part of the target. In this sense, I agree renaming `Target` to `ISA` to be more explicit.

Reply:
I was viewing the target more as the model of a backend for the default compiler (ie via preset passmanagers, custom things can still be handled out of band like they are today) to use when compiling an optimized circuit (or pulse schedule) for a backend. The target object right now is basically just what I was calling the `gate_map` in the first iteration of this PR (many months ago) which is basically the same as your ISA idea because right now what the transpiler can use mostly lines up with the just the instruction details, their pulse definitions, etc. But longer term I was thinking it could contain more than just that as the needs of the compiler grow and change (which was what my above comment was about).

I don't have strong opinions on the qubits properties payload or where it lives for this PR, but I agree it should be defined using separate data structure (which right now is modeled as a list of `QubitProperties` objects) as it's a fundamentally different. I was just musing that in a potential future where we start using it in the preset passmanagers or other default compiler configuration at some point we should have a way to include that data structure in the target class. What I'm really trying to do though when this time comes to expand things like this though is doing it in a controlled fashion instead of adding ad-hoc separate fields and data structures to the configuration payload like we have been. I was thinking this would be done via the target either via versioning, adding mixins (like a `TargetQubitProperties` mixin), or subclassing to add additional attributes to a Target. But I think this is something we can discuss as a follow up to this when the time comes as it's really a separate discussion. I think for this PR and the immediate future we can view the target as closely matching your ISA tagged issues, and it gets us so much more flexibility for the compiler that expanding on this won't really come up for a while. (mostly because there is the long tail after this to update the transpiler passes to work with a target directly which is a huge chunk of work). I don't want to bog down this work on discussions of how we evolve things down the road, we can do that later (which is really what a rfc/design doc discussion would be useful for) after this is in place.

> If we need to pack everything in Target, I wonder how we can set a scheduling option ("alap"/"asap"/None) as a part of the target.

I don't think we want to put everything in the target, the target is just the object that models a backend for the compiler. `alap/asap/None` for the scheduling method is an argument/option for the compiler. The simplest way I think to look at is in the code: https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/transpiler/passmanager_config.py#L106-L115 which initializes a new `PassManagerConfig` (which is the object we pass to the preset passmanager constructors) from a backend. The target I view as the object replacing the highlighted lines (substitute in `backend.target` and whatever those would grow to over time as the needs of the compiler changed. The user configurable options are still separate from that.

> In this sense, I agree renaming Target to ISA to be more explicit.

I actually don't think ISA is an accurate name for this data structure because this is more than just defining the instruction set architecture. If it was just the basis gates and the coupling map/interaction map I'd view that as more of an isa. But we start adding the tuning parameters like duration, error rates, and pulse calibrations which differ to QPU to QPU (even if they are the same architecture and use the same instructions and have the same connectivity) or over time that really is a closer model to a CPU specific compiler target than an isa. For example, setting llvm to target `x86_64` vs `znver2` the first is an abstract target for the generic x86_64 isa while the second is a target tuned specifically for my [workstation's CPU](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/X86/X86ScheduleZnver2.td) that includes specific characteristics of the hardware (like cycle latency for instructions, timing characteristics, etc) for the AMD Zen version 2 implementation of x86_64. This distinction actually (very loosely) is what I was modelling the target class on.